File: sdk/js/gladys-sensor-sdk/test/EmitBatch.test.ts
import { afterEach, describe, expect, it, vi } from "vitest";
import { EventBuilder } from "../src/EventBuilder";
import { EventDispatcher, type EmitResult } from "../src/EventDispatcher";
import { FlowStrategy, RateLimitStrategy } from "../src/FlowStrategy";
import { Event } from "../src/generated/common";

type MockClient = {
  publishEvent: ReturnType<typeof vi.fn>;
  publishEvents: ReturnType<typeof vi.fn>;
};

class BudgetStrategy implements FlowStrategy {
  private budget: number;
  availableCalls = 0;
  consumeCalls: number[] = [];

  constructor(budget: number) {
    this.budget = budget;
  }

  shouldPublish(_event: Event): boolean {
    return true;
  }

  availableTokens(): number {
    this.availableCalls += 1;
    return this.budget;
  }

  consume(n: number): void {
    this.consumeCalls.push(n);
    this.budget -= n;
  }
}

function makeClient(): MockClient {
  return {
    publishEvent: vi.fn().mockResolvedValue(undefined),
    publishEvents: vi.fn().mockResolvedValue(undefined),
  };
}

function makeEvent(
  label: string,
  options: { threat?: boolean; priority?: number } = {}
): Event {
  const event = new EventBuilder("sensor")
    .text(label)
    .structured({ priority: options.priority ?? 0 })
    .build();
  if (options.threat) {
    return new EventBuilder("sensor")
      .text(label)
      .structured({ priority: options.priority ?? 0 })
      .threat(true)
      .build();
  }
  return event;
}

function sentLabels(client: MockClient): string[] {
  const labels: string[] = [];
  for (const call of client.publishEvent.mock.calls) {
    labels.push((call[0] as Event).rawText ?? "");
  }
  for (const call of client.publishEvents.mock.calls) {
    labels.push(...(call[0] as Event[]).map((event) => event.rawText ?? ""));
  }
  return labels;
}

function expectResult(result: EmitResult, sent: number, suppressed: number): void {
  expect(result).toEqual({ sent, suppressed });
}

describe("EmitBatch", () => {
  afterEach(() => {
    vi.restoreAllMocks();
  });

  it("test_emit_batch_empty_list", async () => {
    const strategy = new BudgetStrategy(10);
    const client = makeClient();
    const dispatcher = new EventDispatcher(client as any, "sensor", { strategy });

    const result = await dispatcher.emitBatch([]);

    expectResult(result, 0, 0);
    expect(strategy.availableCalls).toBe(0);
    expect(strategy.consumeCalls).toEqual([]);
    expect(client.publishEvent).not.toHaveBeenCalled();
    expect(client.publishEvents).not.toHaveBeenCalled();
  });

  it("test_emit_batch_all_threats", async () => {
    const strategy = new BudgetStrategy(0);
    const client = makeClient();
    const dispatcher = new EventDispatcher(client as any, "sensor", { strategy });

    const result = await dispatcher.emitBatch([
      makeEvent("t1", { threat: true }),
      makeEvent("t2", { threat: true }),
    ]);

    expectResult(result, 2, 0);
    expect(strategy.availableCalls).toBe(0);
    expect(strategy.consumeCalls).toEqual([]);
    expect(sentLabels(client)).toEqual(["t1", "t2"]);
  });

  it("test_emit_batch_all_within_budget", async () => {
    const strategy = new BudgetStrategy(3);
    const client = makeClient();
    const dispatcher = new EventDispatcher(client as any, "sensor", { strategy });

    const result = await dispatcher.emitBatch([
      makeEvent("a"),
      makeEvent("b"),
      makeEvent("c"),
    ]);

    expectResult(result, 3, 0);
    expect(strategy.consumeCalls).toEqual([3]);
    expect(sentLabels(client)).toEqual(["a", "b", "c"]);
  });

  it("test_emit_batch_zero_budget", async () => {
    const strategy = new BudgetStrategy(0);
    const client = makeClient();
    const dispatcher = new EventDispatcher(client as any, "sensor", { strategy });

    const result = await dispatcher.emitBatch([
      makeEvent("t", { threat: true }),
      makeEvent("a"),
      makeEvent("b"),
    ]);

    expectResult(result, 1, 2);
    expect(strategy.consumeCalls).toEqual([]);
    expect(dispatcher.eventsFiltered).toBe(2);
    expect(sentLabels(client)).toEqual(["t"]);
  });

  it("test_emit_batch_single_event", async () => {
    const strategy = new BudgetStrategy(1);
    const client = makeClient();
    const dispatcher = new EventDispatcher(client as any, "sensor", { strategy });

    const result = await dispatcher.emitBatch([makeEvent("single")]);

    expectResult(result, 1, 0);
    expect(strategy.consumeCalls).toEqual([1]);
    expect(client.publishEvent).toHaveBeenCalledTimes(1);
  });

  it("test_emit_batch_fifo_when_no_priority_fn", async () => {
    const strategy = new BudgetStrategy(2);
    const client = makeClient();
    const dispatcher = new EventDispatcher(client as any, "sensor", { strategy });

    const result = await dispatcher.emitBatch([
      makeEvent("a"),
      makeEvent("b"),
      makeEvent("c"),
    ]);

    expectResult(result, 2, 1);
    expect(sentLabels(client)).toEqual(["a", "b"]);
  });

  it("test_emit_batch_priority_fn_selects_top_n", async () => {
    const strategy = new BudgetStrategy(2);
    const client = makeClient();
    const dispatcher = new EventDispatcher(client as any, "sensor", {
      strategy,
      priorityFn: (event) => Number((event.structured as any)?.priority ?? 0),
    });

    const result = await dispatcher.emitBatch([
      makeEvent("a", { priority: 1 }),
      makeEvent("b", { priority: 10 }),
      makeEvent("c", { priority: 5 }),
    ]);

    expectResult(result, 2, 1);
    expect(sentLabels(client)).toEqual(["b", "c"]);
  });

  it("test_emit_batch_priority_fn_preserves_order", async () => {
    const strategy = new BudgetStrategy(2);
    const client = makeClient();
    const dispatcher = new EventDispatcher(client as any, "sensor", {
      strategy,
      priorityFn: (event) => Number((event.structured as any)?.priority ?? 0),
    });

    const result = await dispatcher.emitBatch([
      makeEvent("a", { priority: 5 }),
      makeEvent("b", { priority: 1 }),
      makeEvent("c", { priority: 10 }),
    ]);

    expectResult(result, 2, 1);
    expect(sentLabels(client)).toEqual(["a", "c"]);
  });

  it("test_emit_batch_equal_priority_preserves_order", async () => {
    const strategy = new BudgetStrategy(2);
    const client = makeClient();
    const dispatcher = new EventDispatcher(client as any, "sensor", {
      strategy,
      priorityFn: () => 1,
    });

    const result = await dispatcher.emitBatch([
      makeEvent("a"),
      makeEvent("b"),
      makeEvent("c"),
    ]);

    expectResult(result, 2, 1);
    expect(sentLabels(client)).toEqual(["a", "b"]);
  });

  it("test_emit_batch_threats_bypass_budget", async () => {
    const strategy = new BudgetStrategy(0);
    const client = makeClient();
    const dispatcher = new EventDispatcher(client as any, "sensor", { strategy });

    const result = await dispatcher.emitBatch([
      makeEvent("t1", { threat: true }),
      makeEvent("a"),
      makeEvent("t2", { threat: true }),
      makeEvent("b"),
    ]);

    expectResult(result, 2, 2);
    expect(sentLabels(client)).toEqual(["t1", "t2"]);
  });

  it("test_emit_batch_threats_dont_consume_tokens", async () => {
    vi.spyOn(Date, "now").mockReturnValue(0);
    const strategy = new RateLimitStrategy(5, 10);
    const client = makeClient();
    const dispatcher = new EventDispatcher(client as any, "sensor", { strategy });
    const events = Array.from({ length: 10 }, (_, i) =>
      makeEvent(`t${i}`, { threat: true })
    );

    const before = strategy.availableTokens();
    const result = await dispatcher.emitBatch(events);
    const after = strategy.availableTokens();

    expectResult(result, 10, 0);
    expect(before).toBe(5);
    expect(after).toBe(5);
  });

  it("test_emit_batch_mixed_threats_and_candidates", async () => {
    const strategy = new BudgetStrategy(1);
    const client = makeClient();
    const dispatcher = new EventDispatcher(client as any, "sensor", { strategy });

    const result = await dispatcher.emitBatch([
      makeEvent("t1", { threat: true }),
      makeEvent("a"),
      makeEvent("b"),
      makeEvent("t2", { threat: true }),
    ]);

    expectResult(result, 3, 1);
    expect(sentLabels(client)).toEqual(["t1", "a", "t2"]);
  });

  it("test_emit_batch_updates_events_filtered", async () => {
    const strategy = new BudgetStrategy(1);
    const client = makeClient();
    const dispatcher = new EventDispatcher(client as any, "sensor", { strategy });

    const result = await dispatcher.emitBatch([
      makeEvent("a"),
      makeEvent("b"),
      makeEvent("c"),
    ]);

    expectResult(result, 1, 2);
    expect(dispatcher.eventsFiltered).toBe(2);
  });

  it("test_emit_batch_updates_events_published", async () => {
    const strategy = new BudgetStrategy(1);
    const client = makeClient();
    const dispatcher = new EventDispatcher(client as any, "sensor", { strategy });

    const result = await dispatcher.emitBatch([
      makeEvent("t", { threat: true }),
      makeEvent("a"),
      makeEvent("b"),
    ]);

    expectResult(result, 2, 1);
    expect(dispatcher.eventsPublished).toBe(2);
  });

  it("test_emit_result_sent_plus_suppressed_equals_total", async () => {
    const scenarios: Array<{ budget: number; events: Event[] }> = [
      { budget: 0, events: [] },
      { budget: 0, events: [makeEvent("a"), makeEvent("b"), makeEvent("t", { threat: true })] },
      { budget: 2, events: [makeEvent("a"), makeEvent("b"), makeEvent("c")] },
      { budget: 1, events: [makeEvent("t", { threat: true }), makeEvent("a"), makeEvent("b")] },
    ];

    for (const scenario of scenarios) {
      const strategy = new BudgetStrategy(scenario.budget);
      const client = makeClient();
      const dispatcher = new EventDispatcher(client as any, "sensor", { strategy });
      const result = await dispatcher.emitBatch(scenario.events);
      expect(result.sent + result.suppressed).toBe(scenario.events.length);
    }
  });
});

File: sdk/java/gladys-sensor-sdk/src/test/java/com/gladys/sensor/EmitBatchTest.java
package com.gladys.sensor;

import com.google.protobuf.Value;
import gladys.types.Types;
import gladys.v1.Common;
import gladys.v1.Orchestrator;
import gladys.v1.OrchestratorServiceGrpc;
import io.grpc.ManagedChannel;
import io.grpc.Server;
import io.grpc.inprocess.InProcessChannelBuilder;
import io.grpc.inprocess.InProcessServerBuilder;
import io.grpc.stub.StreamObserver;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.concurrent.atomic.AtomicLong;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;

class EmitBatchTest {
    private final AtomicInteger publishEventCount = new AtomicInteger();
    private final List<Common.Event> publishedEvents = new ArrayList<>();
    private Server server;
    private ManagedChannel channel;
    private GladysClient client;

    private final OrchestratorServiceGrpc.OrchestratorServiceImplBase serviceImpl =
            new OrchestratorServiceGrpc.OrchestratorServiceImplBase() {
                @Override
                public void publishEvent(Orchestrator.PublishEventRequest request,
                                         StreamObserver<Orchestrator.PublishEventResponse> responseObserver) {
                    publishEventCount.incrementAndGet();
                    publishedEvents.add(request.getEvent());
                    responseObserver.onNext(Orchestrator.PublishEventResponse.newBuilder()
                            .setAck(Orchestrator.EventAck.newBuilder()
                                    .setEventId(request.getEvent().getId())
                                    .setAccepted(true)
                                    .build())
                            .build());
                    responseObserver.onCompleted();
                }

                @Override
                public void publishEvents(Orchestrator.PublishEventsRequest request,
                                          StreamObserver<Orchestrator.PublishEventsResponse> responseObserver) {
                    publishEventCount.addAndGet(request.getEventsCount());
                    publishedEvents.addAll(request.getEventsList());
                    responseObserver.onNext(Orchestrator.PublishEventsResponse.newBuilder()
                            .setAcceptedCount(request.getEventsCount())
                            .build());
                    responseObserver.onCompleted();
                }
            };

    @BeforeEach
    void setUp() throws Exception {
        publishEventCount.set(0);
        publishedEvents.clear();
        String serverName = InProcessServerBuilder.generateName();

        server = InProcessServerBuilder.forName(serverName)
                .directExecutor()
                .addService(serviceImpl)
                .build()
                .start();

        channel = InProcessChannelBuilder.forName(serverName)
                .directExecutor()
                .build();

        client = new GladysClient(channel);
    }

    @AfterEach
    void tearDown() {
        client.close();
        server.shutdownNow();
    }

    @Test
    void test_emit_batch_empty_list() {
        BudgetStrategy strategy = new BudgetStrategy(10);
        EventDispatcher dispatcher = new EventDispatcher(client, "sensor", 0, true, strategy);

        EmitResult result = dispatcher.emitBatch(List.of());

        assertEquals(new EmitResult(0, 0), result);
        assertEquals(0, strategy.availableCalls);
        assertEquals(List.of(), strategy.consumeCalls);
        assertEquals(0, publishEventCount.get());
        dispatcher.shutdown();
    }

    @Test
    void test_emit_batch_all_threats() {
        BudgetStrategy strategy = new BudgetStrategy(0);
        EventDispatcher dispatcher = new EventDispatcher(client, "sensor", 0, true, strategy);

        EmitResult result = dispatcher.emitBatch(List.of(
                threatEvent("t1"),
                threatEvent("t2")
        ));

        assertEquals(new EmitResult(2, 0), result);
        assertEquals(0, strategy.availableCalls);
        assertEquals(List.of(), strategy.consumeCalls);
        assertEquals(List.of("t1", "t2"), labels());
        dispatcher.shutdown();
    }

    @Test
    void test_emit_batch_all_within_budget() {
        BudgetStrategy strategy = new BudgetStrategy(3);
        EventDispatcher dispatcher = new EventDispatcher(client, "sensor", 0, true, strategy);

        EmitResult result = dispatcher.emitBatch(List.of(
                event("a", 1),
                event("b", 2),
                event("c", 3)
        ));

        assertEquals(new EmitResult(3, 0), result);
        assertEquals(List.of(3), strategy.consumeCalls);
        assertEquals(List.of("a", "b", "c"), labels());
        dispatcher.shutdown();
    }

    @Test
    void test_emit_batch_zero_budget() {
        BudgetStrategy strategy = new BudgetStrategy(0);
        EventDispatcher dispatcher = new EventDispatcher(client, "sensor", 0, true, strategy);

        EmitResult result = dispatcher.emitBatch(List.of(
                threatEvent("t"),
                event("a", 1),
                event("b", 2)
        ));

        assertEquals(new EmitResult(1, 2), result);
        assertEquals(2, dispatcher.getEventsFiltered());
        assertEquals(List.of(), strategy.consumeCalls);
        assertEquals(List.of("t"), labels());
        dispatcher.shutdown();
    }

    @Test
    void test_emit_batch_single_event() {
        BudgetStrategy strategy = new BudgetStrategy(1);
        EventDispatcher dispatcher = new EventDispatcher(client, "sensor", 0, true, strategy);

        EmitResult result = dispatcher.emitBatch(List.of(event("single", 1)));

        assertEquals(new EmitResult(1, 0), result);
        assertEquals(List.of(1), strategy.consumeCalls);
        assertEquals(1, publishEventCount.get());
        dispatcher.shutdown();
    }

    @Test
    void test_emit_batch_fifo_when_no_priority_fn() {
        BudgetStrategy strategy = new BudgetStrategy(2);
        EventDispatcher dispatcher = new EventDispatcher(client, "sensor", 0, true, strategy);

        EmitResult result = dispatcher.emitBatch(List.of(
                event("a", 1),
                event("b", 2),
                event("c", 3)
        ));

        assertEquals(new EmitResult(2, 1), result);
        assertEquals(List.of("a", "b"), labels());
        dispatcher.shutdown();
    }

    @Test
    void test_emit_batch_priority_fn_selects_top_n() {
        BudgetStrategy strategy = new BudgetStrategy(2);
        EventDispatcher dispatcher = new EventDispatcher(
                client,
                "sensor",
                0,
                true,
                strategy,
                this::priority
        );

        EmitResult result = dispatcher.emitBatch(List.of(
                event("a", 1),
                event("b", 10),
                event("c", 5)
        ));

        assertEquals(new EmitResult(2, 1), result);
        assertEquals(List.of("b", "c"), labels());
        dispatcher.shutdown();
    }

    @Test
    void test_emit_batch_priority_fn_preserves_order() {
        BudgetStrategy strategy = new BudgetStrategy(2);
        EventDispatcher dispatcher = new EventDispatcher(
                client,
                "sensor",
                0,
                true,
                strategy,
                this::priority
        );

        EmitResult result = dispatcher.emitBatch(List.of(
                event("a", 5),
                event("b", 1),
                event("c", 10)
        ));

        assertEquals(new EmitResult(2, 1), result);
        assertEquals(List.of("a", "c"), labels());
        dispatcher.shutdown();
    }

    @Test
    void test_emit_batch_equal_priority_preserves_order() {
        BudgetStrategy strategy = new BudgetStrategy(2);
        EventDispatcher dispatcher = new EventDispatcher(
                client,
                "sensor",
                0,
                true,
                strategy,
                event -> 1.0
        );

        EmitResult result = dispatcher.emitBatch(List.of(
                event("a", 1),
                event("b", 1),
                event("c", 1)
        ));

        assertEquals(new EmitResult(2, 1), result);
        assertEquals(List.of("a", "b"), labels());
        dispatcher.shutdown();
    }

    @Test
    void test_emit_batch_threats_bypass_budget() {
        BudgetStrategy strategy = new BudgetStrategy(0);
        EventDispatcher dispatcher = new EventDispatcher(client, "sensor", 0, true, strategy);

        EmitResult result = dispatcher.emitBatch(List.of(
                threatEvent("t1"),
                event("a", 1),
                threatEvent("t2"),
                event("b", 2)
        ));

        assertEquals(new EmitResult(2, 2), result);
        assertEquals(List.of("t1", "t2"), labels());
        dispatcher.shutdown();
    }

    @Test
    void test_emit_batch_threats_dont_consume_tokens() {
        AtomicLong now = new AtomicLong(0L);
        RateLimitStrategy strategy = new RateLimitStrategy(5, 10, now::get);
        EventDispatcher dispatcher = new EventDispatcher(client, "sensor", 0, true, strategy);
        List<Common.Event> events = new ArrayList<>();
        for (int i = 0; i < 10; i++) {
            events.add(threatEvent("t" + i));
        }

        int before = strategy.availableTokens();
        EmitResult result = dispatcher.emitBatch(events);
        int after = strategy.availableTokens();

        assertEquals(new EmitResult(10, 0), result);
        assertEquals(5, before);
        assertEquals(5, after);
        dispatcher.shutdown();
    }

    @Test
    void test_emit_batch_mixed_threats_and_candidates() {
        BudgetStrategy strategy = new BudgetStrategy(1);
        EventDispatcher dispatcher = new EventDispatcher(client, "sensor", 0, true, strategy);

        EmitResult result = dispatcher.emitBatch(List.of(
                threatEvent("t1"),
                event("a", 1),
                event("b", 2),
                threatEvent("t2")
        ));

        assertEquals(new EmitResult(3, 1), result);
        assertEquals(List.of("t1", "a", "t2"), labels());
        dispatcher.shutdown();
    }

    @Test
    void test_emit_batch_updates_events_filtered() {
        BudgetStrategy strategy = new BudgetStrategy(1);
        EventDispatcher dispatcher = new EventDispatcher(client, "sensor", 0, true, strategy);

        EmitResult result = dispatcher.emitBatch(List.of(
                event("a", 1),
                event("b", 2),
                event("c", 3)
        ));

        assertEquals(new EmitResult(1, 2), result);
        assertEquals(2, dispatcher.getEventsFiltered());
        dispatcher.shutdown();
    }

    @Test
    void test_emit_batch_updates_events_published() {
        BudgetStrategy strategy = new BudgetStrategy(1);
        EventDispatcher dispatcher = new EventDispatcher(client, "sensor", 0, true, strategy);

        EmitResult result = dispatcher.emitBatch(List.of(
                threatEvent("t"),
                event("a", 1),
                event("b", 2)
        ));

        assertEquals(new EmitResult(2, 1), result);
        assertEquals(2, dispatcher.getEventsPublished());
        dispatcher.shutdown();
    }

    @Test
    void test_emit_result_sent_plus_suppressed_equals_total() {
        List<Scenario> scenarios = List.of(
                new Scenario(0, List.of()),
                new Scenario(0, List.of(event("a", 1), event("b", 2), threatEvent("t"))),
                new Scenario(2, List.of(event("a", 1), event("b", 2), event("c", 3))),
                new Scenario(1, List.of(threatEvent("t"), event("a", 1), event("b", 2)))
        );

        for (Scenario scenario : scenarios) {
            BudgetStrategy strategy = new BudgetStrategy(scenario.budget);
            EventDispatcher dispatcher = new EventDispatcher(client, "sensor", 0, true, strategy);
            EmitResult result = dispatcher.emitBatch(scenario.events);
            assertEquals(scenario.events.size(), result.sent() + result.suppressed());
            dispatcher.shutdown();
        }
    }

    private Common.Event event(String label, double priority) {
        return new EventBuilder("sensor")
                .text(label)
                .structured(Map.of("priority", priority))
                .build();
    }

    private Common.Event threatEvent(String label) {
        return event(label, 0).toBuilder()
                .setSalience(Types.SalienceResult.newBuilder().setThreat(1.0f).build())
                .build();
    }

    private double priority(Common.Event event) {
        Value value = event.getStructured().getFieldsMap().get("priority");
        if (value == null) {
            return 0;
        }
        return value.getNumberValue();
    }

    private List<String> labels() {
        List<String> labels = new ArrayList<>();
        for (Common.Event event : publishedEvents) {
            labels.add(event.getRawText());
        }
        return labels;
    }

    private static final class BudgetStrategy implements FlowStrategy {
        private int budget;
        private int availableCalls = 0;
        private final List<Integer> consumeCalls = new ArrayList<>();

        private BudgetStrategy(int budget) {
            this.budget = budget;
        }

        @Override
        public boolean shouldPublish(Common.Event event) {
            return true;
        }

        @Override
        public int availableTokens() {
            availableCalls += 1;
            return budget;
        }

        @Override
        public void consume(int n) {
            consumeCalls.add(n);
            budget -= n;
        }
    }

    private static final class Scenario {
        private final int budget;
        private final List<Common.Event> events;

        private Scenario(int budget, List<Common.Event> events) {
            this.budget = budget;
            this.events = events;
        }
    }
}
